File: src/main/java/com/matchilling/playground/java/Recursion.java
package com.matchilling.playground.java;

public class Recursion {

    public static void countDown(int number) {
        if (number > 0) {
            System.out.println(number);
            countDown(number - 1);
        }
    }

    public static void countUp(int number) {
        countUpRecursive(number, 0);
    }

    private static void countUpRecursive(int number, int counter) {
        if (counter < number) {
            System.out.println(++counter);
            countUpRecursive(number, counter);
        }
    }
}

File: src/main/scala/com/matchilling/playground/scala/Recursion.scala
package com.matchilling.playground.scala

import scala.annotation.tailrec

object Recursion {

  @tailrec
  def countDown(number: Int): Unit = {
    if (number > 0) {
      println(number)
      countDown(number - 1)
    }
  }

  def countUp(number: Int): Unit = {
    if (number > 0) {
      countUp(number - 1)
      println(number)
    }
  }

  def factorialNotTailOptimised(number: Int): Int = {
    if (number == 0) return 1
    number * factorialNotTailOptimised(number - 1)
  }

  def factorial(number: Int): Int = {
    @tailrec
    def iterator(number: Int, result: Int): Int = {
      if (number == 0) return result
      iterator(number - 1, result * number)
    }

    iterator(number, 1)
  }
}
